backend/api/Repositories/Cdogs/CdogsAuthRepository.cs

```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Pims.Api.Helpers.Exceptions;
+using Pims.Api.Models;
+using Pims.Api.Models.Cdogs;
+
+namespace Pims.Api.Repositories.Cdogs
+{
+    /// <summary>
+    /// CdogsAuthRepository provides authentication to the Cdogs external api.
+    /// </summary>
+    public class CdogsAuthRepository : CdogsBaseRepository, IDocumentGenerationAuthRepository
+    {
+        private JwtResponse _currentToken;
+        private DateTime _lastSucessfullRequest;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CdogsAuthRepository"/> class.
+        /// </summary>
+        /// <param name="logger">Injected Logger Provider.</param>
+        /// <param name="httpClientFactory">Injected Httpclient factory.</param>
+        /// <param name="configuration">The injected configuration provider.</param>
+        public CdogsAuthRepository(
+            ILogger<CdogsAuthRepository> logger,
+            IHttpClientFactory httpClientFactory,
+            IConfiguration configuration)
+            : base(logger, httpClientFactory, configuration)
+        {
+            _currentToken = null;
+            _lastSucessfullRequest = DateTime.UnixEpoch;
+        }
+
+        public async Task<string> GetTokenAsync()
+        {
+            if (!IsValidToken())
+            {
+                ExternalResult<JwtResponse> tokenResult = await RequestToken();
+
+                if (tokenResult.Status == ExternalResultStatus.Error)
+                {
+                    throw new AuthenticationException(tokenResult.Message);
+                }
+
+                _lastSucessfullRequest = DateTime.UtcNow;
+                _currentToken = tokenResult.Payload;
+            }
+
+            return _currentToken.AccessToken;
+        }
+
+        private bool IsValidToken()
+        {
+            if (_currentToken != null)
+            {
+                DateTime now = DateTime.UtcNow;
+                TimeSpan delta = now - _lastSucessfullRequest;
+                if (delta.TotalSeconds >= _currentToken.ExpiresIn)
+                {
+                    // Revoke token
+                    _logger.LogDebug("Authentication Token has expired.");
+                    _currentToken = null;
+                    return false;
+                }
+                return true;
+            }
+
+            return false;
+        }
+
+        private async Task<ExternalResult<JwtResponse>> RequestToken()
+        {
+            _logger.LogDebug("Getting authentication token...");
+
+            using HttpClient client = _httpClientFactory.CreateClient("Pims.Api.Logging");
+            client.DefaultRequestHeaders.Accept.Clear();
+
+            Uri endpoint = new Uri(_config.AuthHost, "auth/realms/jbd6rnxw/protocol/openid-connect/token");
```

Comment:
I would think that the realm needs to be .env configurable. There are dev/test/prod instances of cdogs so I would assume there are dev/test/prod keycloak realms as well.